cli/flox-watchdog/src/logger.rs

```diff
@@ -61,17 +61,21 @@ pub(crate) fn spawn_heartbeat_log() {
 
 /// Starts a background thread which garbage collects known log files. This is
 /// done on a best effort basis; errors are traced rather than being bubbled up
-/// and the thread will run until the watchdog exits.
+/// and the thread will loop until the watchdog exits.
+///
+/// All of the functions called here must be deterministic because there may be
+/// multiple watchdogs running for the same environment log dir.
 pub(crate) fn spawn_logs_gc_threads(dir: impl AsRef<Path>) {
     let dir = dir.as_ref().to_path_buf();
-    let dir_clone = dir.clone();
-    spawn(move || {
-        gc_logs_watchdog(dir_clone, KEEP_WATCHDOG_DAYS)
+    spawn(move || loop {
+        gc_logs_watchdog(&dir, KEEP_WATCHDOG_DAYS)
             .unwrap_or_else(|err| error!(%err, "failed to delete watchdog logs"));
-    });
-    spawn(move || {
-        gc_logs_services(&dir, KEEP_SERVICES_LAST)
+        gc_logs_per_process(&dir, "services.*.log", KEEP_LAST_N_PROCESSES)
             .unwrap_or_else(|err| error!(%err, "failed to delete services logs"));
+        gc_logs_per_process(&dir, "upgrade-check.*.log", KEEP_LAST_N_PROCESSES)
+            .unwrap_or_else(|err| error!(%err, "failed to delete upgrade-check logs"));
```

Comment:
Sidenote: it would be nice to share constants for these but they're all over the place. One is defined in bash and the other is in the CLI which the watchdog can't/shouldn't depend on. We could put the latter in `flox-core`? 🤷🏻

Reply:
At a minimum I would share a constant between here and the tests

Reply:
Although actually now I see we're creating the log file name in the test, so it's not actually testing that we cleanup our logs. Unless we factor that out as well I'm not sure it matters

Reply:
To answer the original question, `flox-core` is the place to put constants like this that you want shared between different crates.

Reply:
Extracted:

> Share the construction of the filename pattern between the CLI and
> watchdog, so that they don't get out-of-sync and evade GC.
> 
> This can't be constant because `format!` needs a string literal.
> 
> We can't do the same for services logs because the filename is
> constructed in bash. And newer watchdog logs use a different pattern for
> log rotation.